File: core/src/nl/t64/game/rpg/ProfileManager.java
package nl.t64.game.rpg;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Preferences;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.Json;
import com.badlogic.gdx.utils.ObjectMap;
import lombok.Setter;
import nl.t64.game.rpg.constants.Constant;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.stream.IntStream;


public class ProfileManager {

    private static final String DEFAULT_ID =
            Constant.PLAYER_ID.substring(0, 1).toUpperCase() + Constant.PLAYER_ID.substring(1);
    private static final String SAVE1 = "save1.dat";
    private static final String SAVE2 = "save2.dat";
    private static final String SAVE3 = "save3.dat";
    private static final String SAVE4 = "save4.dat";
    private static final String SAVE5 = "save5.dat";
    private static final List<String> SAVE_FILES = List.of(SAVE1, SAVE2, SAVE3, SAVE4, SAVE5);
    private static final String SAVE_STATE_KEY = "saveState";
    private static final String PROFILE_ID = "id";
    private static final String PROFILE_SAVE_DATE = "saveDate";
    private static final String DATE_PATTERN = "yyyy-MM-dd HH:mm";

    private final Json json = new Json();
    private ObjectMap<String, Object> saveStateProperties = new ObjectMap<>();
    private String currentSaveFileName = null;
    @Setter
    private int selectedIndex = -1;

    public boolean doesProfileExist(int profileIndex) {
        Preferences saveFile = getSaveFileBy(profileIndex);
        return saveFile.contains(SAVE_STATE_KEY);
    }

    public void createNewProfile(String profileId) {
        if (profileId.isEmpty()) {
            profileId = DEFAULT_ID;
        }
        currentSaveFileName = SAVE_FILES.get(selectedIndex);
        saveStateProperties.clear();
        setProperty(PROFILE_ID, profileId);
        Utils.getBrokerManager().profileObservers.notifyCreateProfile(this);
        writeProfileToDisk();
    }

    public <T> T getProperty(String key, Class<T> clazz) {
        return clazz.cast(saveStateProperties.get(key));
    }

    public void setProperty(String key, Object object) {
        saveStateProperties.put(key, object);
    }

    public void saveProfile() {
        Utils.getBrokerManager().profileObservers.notifySaveProfile(this);
        writeProfileToDisk();
    }

    public void loadProfile(int profileIndex) {
        currentSaveFileName = SAVE_FILES.get(profileIndex);
        Preferences saveFile = Gdx.app.getPreferences(currentSaveFileName);
        saveStateProperties = getSaveStateProperties(saveFile);
        Utils.getBrokerManager().profileObservers.notifyLoadProfile(this);
    }

    public void removeProfile(int profileIndex) {
        Preferences saveFile = getSaveFileBy(profileIndex);
        saveFile.clear();
        saveFile.flush();
    }

    public Array<String> getVisualProfileArray() {
        String[] visualListOfProfiles = IntStream.range(0, SAVE_FILES.size())
                                                 .mapToObj(this::getVisualOf)
                                                 .toArray(String[]::new);
        return new Array<>(visualListOfProfiles);
    }

    private void writeProfileToDisk() {
        var formatter = DateTimeFormatter.ofPattern(DATE_PATTERN);
        setProperty(PROFILE_SAVE_DATE, formatter.format(LocalDateTime.now()));
        String saveStateJsonString = json.prettyPrint(json.toJson(saveStateProperties));
        Preferences saveFile = Gdx.app.getPreferences(currentSaveFileName);
        saveFile.putString(SAVE_STATE_KEY, saveStateJsonString);
        saveFile.flush();
    }

    private String getVisualOf(int profileIndex) {
        Preferences saveFile = getSaveFileBy(profileIndex);
        if (saveFile.contains(SAVE_STATE_KEY)) {
            ObjectMap<String, Object> objectMap = getSaveStateProperties(saveFile);
            return objectMap.get(PROFILE_ID) + " [" + objectMap.get(PROFILE_SAVE_DATE) + "]";
        } else {
            return (profileIndex + 1) + " [...]";
        }
    }

    private Preferences getSaveFileBy(int profileIndex) {
        String saveFileName = SAVE_FILES.get(profileIndex);
        return Gdx.app.getPreferences(saveFileName);
    }

    private ObjectMap<String, Object> getSaveStateProperties(Preferences saveFile) {
        String saveStateJsonString = saveFile.getString(SAVE_STATE_KEY);
        return json.fromJson(ObjectMap.class, saveStateJsonString);
    }

}

File: core/src/nl/t64/game/rpg/ProfileManager.kt
package nl.t64.game.rpg

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Preferences
import com.badlogic.gdx.utils.Json
import ktx.collections.GdxArray
import ktx.collections.GdxMap
import ktx.collections.set
import ktx.json.fromJson
import nl.t64.game.rpg.Utils.brokerManager
import nl.t64.game.rpg.constants.Constant
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter


private const val SAVE_FILES = "save1.dat,save2.dat,save3.dat,save4.dat,save5.dat,save6.dat,save7.dat"
private const val LOADING = ",,Loading...,,,,"
private const val SAVE_STATE_KEY = "saveState"
private const val PROFILE_ID = "id"
private const val PROFILE_SAVE_DATE = "saveDate"
private const val DATE_PATTERN = "yyyy-MM-dd HH:mm"
private const val DEFAULT_EMPTY_PROFILE_VIEW = " [...]"

class ProfileManager {

    private val json = Json()
    private var saveStateProperties = GdxMap<String, Any>()
    private lateinit var currentSaveFileName: String
    var selectedIndex = -1

    fun doesProfileExist(profileIndex: Int): Boolean {
        val saveFile = getSaveFileBy(profileIndex)
        return saveFile.contains(SAVE_STATE_KEY)
    }

    fun createNewProfile(profileId: String) {
        currentSaveFileName = getSaveFileNames()[selectedIndex]
        saveStateProperties.clear()
        brokerManager.profileObservers.notifyCreateProfile(this)
        writeProfileToDisk(profileId.ifEmpty { getDefaultId() })
    }

    @Suppress("NON_PUBLIC_CALL_FROM_PUBLIC_INLINE")
    inline fun <reified T> getProperty(key: String): T {
        return saveStateProperties[key] as T
    }

    fun setProperty(key: String, any: Any) {
        saveStateProperties[key] = any
    }

    fun saveProfile() {
        brokerManager.profileObservers.notifySaveProfile(this)
        writeProfileToDisk(null)
    }

    fun loadProfile(profileIndex: Int) {
        currentSaveFileName = getSaveFileNames()[profileIndex]
        val saveFile = Gdx.app.getPreferences(currentSaveFileName)
        saveStateProperties = getSaveStateProperties(saveFile)
        brokerManager.profileObservers.notifyLoadProfile(this)
    }

    fun removeProfile(profileIndex: Int) {
        val saveFile = getSaveFileBy(profileIndex)
        saveFile.clear()
        saveFile.flush()
    }

    fun getVisualLoadingArray(): GdxArray<String> =
        GdxArray(LOADING.split(",").toTypedArray())

    fun getVisualProfileArray(): GdxArray<String> =
        GdxArray(getSaveFileNames().indices.map { getVisualOf(it) }.toTypedArray())

    fun getLastSaveLocation(): String {
        return saveStateProperties["mapTitle"] as String
    }

    private fun writeProfileToDisk(profileId: String?) {
        val saveDate: String = DateTimeFormatter.ofPattern(DATE_PATTERN).format(LocalDateTime.now())
        val saveStateJsonString = json.prettyPrint(json.toJson(saveStateProperties))
        val saveFile = Gdx.app.getPreferences(currentSaveFileName)
        if (profileId != null) saveFile.putString(PROFILE_ID, profileId)
        saveFile.putString(PROFILE_SAVE_DATE, saveDate)
        saveFile.putString(SAVE_STATE_KEY, saveStateJsonString)
        saveFile.flush()
    }

    private fun getVisualOf(profileIndex: Int): String {
        val saveFile = getSaveFileBy(profileIndex)
        return if (saveFile.contains(SAVE_STATE_KEY)) {
            "${saveFile.getString(PROFILE_ID)} [${saveFile.getString(PROFILE_SAVE_DATE)}]"
        } else {
            "${profileIndex + 1} $DEFAULT_EMPTY_PROFILE_VIEW"
        }
    }

    private fun getSaveFileBy(profileIndex: Int): Preferences {
        val saveFileName = getSaveFileNames()[profileIndex]
        return Gdx.app.getPreferences(saveFileName)
    }

    private fun getSaveStateProperties(saveFile: Preferences): GdxMap<String, Any> {
        val saveStateJsonString = saveFile.getString(SAVE_STATE_KEY)
        return json.fromJson(saveStateJsonString)
    }

    private fun getDefaultId(): String {
        return Constant.PLAYER_ID.replaceFirstChar { it.uppercase() }
    }

    private fun getSaveFileNames(): List<String> = SAVE_FILES.split(",")

}
